File: src/main/java/net/pterodactylus/fcp/quelaton/ClientGetCommand.java
package net.pterodactylus.fcp.quelaton;

import java.io.InputStream;
import java.util.Optional;
import java.util.function.Consumer;

import net.pterodactylus.fcp.Priority;

/**
 * Command that retrieves data from Freenet.
 *
 * @author <a href="mailto:bombe@pterodactylus.net">David ‘Bombe’ Roden</a>
 */
public interface ClientGetCommand {

	ClientGetCommand onRedirect(Consumer<String> onRedirect);
	ClientGetCommand ignoreDataStore();
	ClientGetCommand dataStoreOnly();
	ClientGetCommand maxSize(long maxSize);
	ClientGetCommand priority(Priority priority);
	ClientGetCommand realTime();
	ClientGetCommand global();

	Executable<Optional<Data>> uri(String uri);

	interface Data {

		String getMimeType();
		long size();
		InputStream getInputStream();

	}

}

File: src/main/kotlin/net/pterodactylus/fcp/quelaton/ClientGetCommand.kt
package net.pterodactylus.fcp.quelaton

import net.pterodactylus.fcp.*
import java.io.*
import java.util.*
import java.util.function.*

/**
 * Command that retrieves data from Freenet.
 *
 * @author [David ‘Bombe’ Roden](mailto:bombe@pterodactylus.net)
 */
interface ClientGetCommand {

	fun onRedirect(onRedirect: Consumer<String>): ClientGetCommand
	fun ignoreDataStore(): ClientGetCommand
	fun dataStoreOnly(): ClientGetCommand
	fun maxSize(maxSize: Long): ClientGetCommand
	fun priority(priority: Priority): ClientGetCommand
	fun realTime(): ClientGetCommand
	fun global(): ClientGetCommand

	fun uri(uri: String): Executable<Optional<Data>>

	interface Data {

		val mimeType: String
		val inputStream: InputStream
		fun size(): Long

	}

}
